src/Arcus.Security.Secrets.AzureKeyVault/KeyVaultSecretProvider.cs

```diff
@@ -63,18 +77,43 @@ public async Task<string> Get(string secretName)
                     throw new SecretNotFoundException(secretName, keyVaultErrorException);
                 }
 
+                // TODO: are there other HTTP status codes that we can relate to secret exceptions?
                 throw;
             }
         }
 
         private async Task<KeyVaultClient> GetClientAsync()
         {
-            if (_keyVaultClient == null)
+            await LockCreateKeyVaultClient.WaitAsync();
+
+            try
+            {
+                if (_keyVaultClient == null)
+                {
+                    // TODO: why is this factory not returning an interface (the 'Microsoft.Azure.KeyVault.IKeyVaultClient' interface)?
+                    _keyVaultClient = await _authenticator.Authenticate();
+                }
+
+                return _keyVaultClient;
+            }
+            finally
             {
-                _keyVaultClient = await _authenticator.Authenticate();
+                LockCreateKeyVaultClient.Release();
             }
+        }
+
+        private static Task<SecretBundle> ThrottleTooManyRequests(Func<Task<SecretBundle>> getSecret)
```

Comment:
Given this can be re-used for different operations we should consider renaming it to `secretOperation`

Reply:
Yes